server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperation.java

```diff
@@ -0,0 +1,146 @@
+ * <br/>
+ * Noted that if you add a future delayed operation that calls ReplicaManager.appendRecords() in onComplete()
+ * like DelayedJoin, you must be aware that this operation's onExpiration() needs to call actionQueue.tryCompleteAction().
+ */
+public abstract class DelayedOperation extends TimerTask {
+
+    private final AtomicBoolean completed = new AtomicBoolean(false);
+    // Visible for testing
+    final Lock lock;
+
+    public DelayedOperation(long delayMs) {
+        this(delayMs, Optional.empty());
+    }
+
+    public DelayedOperation(long delayMs, Optional<Lock> lockOpt) {
+        super(delayMs);
+        lock = lockOpt.orElse(new ReentrantLock());
+    }
+
+    /*
+     * Force completing the delayed operation, if not already completed.
+     * This function can be triggered when
+     *
+     * 1. The operation has been verified to be completable inside tryComplete()
+     * 2. The operation has expired and hence needs to be completed right now
+     *
+     * Return true iff the operation is completed by the caller: note that
+     * concurrent threads can try to complete the same operation, but only
+     * the first thread will succeed in completing the operation and return
+     * true, others will still return false
+     */
+    public boolean forceComplete() {
+        if (completed.compareAndSet(false, true)) {
+            // cancel the timeout timer
+            cancel();
+            onComplete();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Check if the delayed operation is already completed
+     */
+    public boolean isCompleted() {
+        return completed.get();
+    }
+
+    /**
+     * Call-back to execute when a delayed operation gets expired and hence forced to complete.
+     */
+    public abstract void onExpiration();
+
+    /**
+     * Process for completing an operation; This function needs to be defined
+     * in subclasses and will be called exactly once in forceComplete()
+     */
+    public abstract void onComplete();
+
+    /**
+     * Try to complete the delayed operation by first checking if the operation
+     * can be completed by now. If yes execute the completion logic by calling
+     * forceComplete() and return true iff forceComplete returns true; otherwise return false
+     * <br/>
+     * This function needs to be defined in subclasses
+     */
+    public abstract boolean tryComplete();
+
+    /**
+     * Thread-safe variant of tryComplete() and call extra function if first tryComplete returns false
+     * @param runnable else function to be executed after first tryComplete returns false
+     * @return result of tryComplete
+     */
+    boolean safeTryCompleteOrElse(Runnable runnable) {
+        lock.lock();
+        try {
+            if (tryComplete()) return true;
+            else {
+                runnable.run();
```

Comment:
Hmm, I think we want the Runnable as input to just invoke method operation as we are not intended to create a new thread here (Runnable is more intended where a new thread is needed). Hence do you think we should create a new functional interface which should be the input for the method?

Reply:
You're right that `Runnable` implies threading. I added a new functional interface.